File: 155-min-stack/min-stack.py
class MinStack:

    def __init__(self):
        self.vals = []
        self.count = -1
        self.min = float('inf')

    def push(self, val: int) -> None:
        self.count += 1
        self.vals.append(val)
        if val < self.min:
            self.min = val

    def pop(self) -> None:
        if self.count >= 0:
            temp = self.vals.pop()
            self.count -= 1
            if temp == self.min:
                self.min = min(self.vals) if self.count >= 0 else float('inf')

    def top(self) -> int:
        if self.count >= 0:
            return self.vals[-1]
        else:
            raise IndexError("Stack is empty")

    def getMin(self) -> int:
        if self.count >= 0:
            return self.min
        else:
            raise IndexError("Stack is empty")


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: 155-min-stack/min-stack.go
type MinStack struct {
    stack []int
    min []int
    top int
}


func Constructor() MinStack {
    return MinStack{[]int{},[]int{},-1}
}


func (this *MinStack) Push(val int)  {
    if this.top == -1 {
        this.min = append(this.min,val)
    } else {
        this.min = append(this.min, min(val,this.min[this.top]))
    }
    this.stack = append(this.stack,val)
    this.top++
}


func (this *MinStack) Pop()  {
    this.stack = this.stack[:this.top]
    this.min = this.min[:this.top]
    this.top--
}


func (this *MinStack) Top() int {
    return this.stack[this.top]
}


func (this *MinStack) GetMin() int {
    return this.min[this.top]
}

func min(a,b int) int {
    if a<b {return a}
    return b
}


/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(val);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.GetMin();
 */